File: src/utils/chessboard.cpp
#include "chess_robot/utils/chessboard.hpp"

#include <opencv2/calib3d.hpp>
#include <opencv2/imgproc.hpp>

namespace chess_robot {

namespace {

constexpr int kClassicBase =
    cv::CALIB_CB_ADAPTIVE_THRESH | cv::CALIB_CB_NORMALIZE_IMAGE;
constexpr int kFastFlags = kClassicBase | cv::CALIB_CB_FAST_CHECK;
constexpr int kSlowFlags = kClassicBase;

std::optional<cv::Mat> findClassic(const cv::Mat& gray, cv::Size pattern_size, int flags) {
  cv::Mat corners;
  const bool found = cv::findChessboardCorners(gray, pattern_size, corners, flags);
  if (!found || corners.empty()) {
    return std::nullopt;
  }
  return refineCorners(gray, corners);
}

std::optional<cv::Mat> findSb(const cv::Mat& gray, cv::Size pattern_size) {
  cv::Mat corners;
  int flags = 0;
#ifdef CALIB_CB_ACCURACY
  flags |= cv::CALIB_CB_ACCURACY;
#endif
  const bool found = cv::findChessboardCornersSB(gray, pattern_size, corners, flags);
  if (!found || corners.empty()) {
    return std::nullopt;
  }
  return corners;
}

}  // namespace

cv::Mat toGray(const cv::Mat& frame) {
  if (frame.channels() == 1) {
    return frame;
  }
  cv::Mat gray;
  cv::cvtColor(frame, gray, cv::COLOR_BGR2GRAY);
  return gray;
}

std::vector<cv::Mat> prepareGrayVariants(const cv::Mat& frame) {
  cv::Mat gray = toGray(frame);
  cv::Ptr<cv::CLAHE> clahe = cv::createCLAHE(2.0, cv::Size(8, 8));
  cv::Mat equalized;
  clahe->apply(gray, equalized);
  cv::Mat blurred;
  cv::GaussianBlur(gray, blurred, cv::Size(5, 5), 0);
  cv::Mat equalized_blurred;
  clahe->apply(blurred, equalized_blurred);
  return {gray, equalized, equalized_blurred};
}

cv::Mat refineCorners(const cv::Mat& gray, const cv::Mat& corners) {
  cv::Mat refined = corners.clone();
  cv::cornerSubPix(
      gray,
      refined,
      cv::Size(11, 11),
      cv::Size(-1, -1),
      cv::TermCriteria(cv::TermCriteria::EPS + cv::TermCriteria::MAX_ITER, 30, 0.001));
  return refined;
}

std::optional<cv::Mat> findInnerCorners(const cv::Mat& frame,
                                        cv::Size pattern_size,
                                        bool robust) {
  if (pattern_size.width < 2 || pattern_size.height < 2) {
    return std::nullopt;
  }

  if (!robust) {
    cv::Mat gray = toGray(frame);
    if (auto corners = findClassic(gray, pattern_size, kFastFlags)) {
      return corners;
    }
    return findClassic(gray, pattern_size, kSlowFlags);
  }

  const std::vector<cv::Mat> variants = prepareGrayVariants(frame);
  for (const cv::Mat& gray : variants) {
    if (auto corners = findClassic(gray, pattern_size, kFastFlags)) {
      return corners;
    }
  }
  for (const cv::Mat& gray : variants) {
    if (auto corners = findClassic(gray, pattern_size, kSlowFlags)) {
      return corners;
    }
    if (auto corners = findSb(gray, pattern_size)) {
      return corners;
    }
  }
  return std::nullopt;
}

}  // namespace chess_robot

File: python/src/utils/chessboard.py
"""Shared OpenCV chessboard / checkerboard inner-corner finding."""

from __future__ import annotations

import cv2
import numpy as np

SUBPIX_CRITERIA = (
    cv2.TERM_CRITERIA_EPS + cv2.TERM_CRITERIA_MAX_ITER,
    30,
    0.001,
)
SUBPIX_WIN_SIZE = (11, 11)
SUBPIX_ZERO_ZONE = (-1, -1)

_CLASSIC_BASE = cv2.CALIB_CB_ADAPTIVE_THRESH | cv2.CALIB_CB_NORMALIZE_IMAGE
_FAST_FLAGS = _CLASSIC_BASE | cv2.CALIB_CB_FAST_CHECK
_SLOW_FLAGS = _CLASSIC_BASE


def to_gray(frame: np.ndarray) -> np.ndarray:
    if frame.ndim == 2:
        return frame
    return cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)


def prepare_gray_variants(frame: np.ndarray) -> list[np.ndarray]:
    """Return gray images with optional CLAHE / blur for harder lighting."""
    gray = to_gray(frame)
    clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
    blurred = cv2.GaussianBlur(gray, (5, 5), 0)
    return [gray, clahe.apply(gray), clahe.apply(blurred)]


def refine_corners(gray: np.ndarray, corners: np.ndarray) -> np.ndarray:
    return cv2.cornerSubPix(
        gray,
        corners,
        winSize=SUBPIX_WIN_SIZE,
        zeroZone=SUBPIX_ZERO_ZONE,
        criteria=SUBPIX_CRITERIA,
    )


def _find_classic(gray: np.ndarray, pattern_size: tuple[int, int], flags: int) -> np.ndarray | None:
    found, corners = cv2.findChessboardCorners(gray, pattern_size, flags)
    if not found or corners is None:
        return None
    return refine_corners(gray, corners)


def _find_sb(gray: np.ndarray, pattern_size: tuple[int, int]) -> np.ndarray | None:
    if not hasattr(cv2, "findChessboardCornersSB"):
        return None

    flags = 0
    if hasattr(cv2, "CALIB_CB_ACCURACY"):
        flags |= cv2.CALIB_CB_ACCURACY

    found, corners = cv2.findChessboardCornersSB(gray, pattern_size, flags=flags)
    if not found or corners is None:
        return None
    return corners


def find_inner_corners(
    frame: np.ndarray,
    pattern_size: tuple[int, int],
    *,
    robust: bool = False,
) -> np.ndarray | None:
    """Find OpenCV inner corners for ``pattern_size`` ``(cols, rows)``.

    ``robust=False`` uses a single gray image (fast, good for calibration HUD).
    ``robust=True`` tries contrast variants, slow classic flags, then SB.
    """
    cols, rows = pattern_size
    if cols < 2 or rows < 2:
        return None

    if not robust:
        gray = to_gray(frame)
        return _find_classic(gray, pattern_size, _FAST_FLAGS) or _find_classic(
            gray,
            pattern_size,
            _SLOW_FLAGS,
        )

    variants = prepare_gray_variants(frame)
    for gray in variants:
        corners = _find_classic(gray, pattern_size, _FAST_FLAGS)
        if corners is not None:
            return corners

    for gray in variants:
        corners = _find_classic(gray, pattern_size, _SLOW_FLAGS)
        if corners is not None:
            return corners
        corners = _find_sb(gray, pattern_size)
        if corners is not None:
            return corners

    return None
